### PR TITLE
bpf: use `bpf_htons` instead of using shift

### PR DESCRIPTION
**Pull request by chez-shanpu**
The current implementation using shift does not take into account endianness.
`bpf_htons()` detects which endianness is used and converts the value appropriately.
Also, this PR defines `bpf_u8_to_be16()` that wraps `bpf_htons()` because converting 8-bit ICMP types to 16-bit does not depend on the host byte order.

resolves https://github.com/cilium/cilium/pull/27369#discussion_r1516958093

### CODE REVIEW

File: bpf/lib/policy.h [hunk 1]
```diff
@@ -69,13 +69,7 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 local_id,
...
 
 # if defined(ENABLE_ICMP_RULE)
-			/* Convert from unsigned char to unsigned short
-			 * considering byte order(little-endian).
-			 * In the little-endian case, for example, 2byte data "AB"
-			 * convert to "BA".
-			 * Therefore, the "icmp_type" should be shifted not just casting.
-			 */
-			key.dport = (__u16)(icmphdr.type << 8);
+			key.dport = bpf_htons(icmphdr.type);
```
Comment (borkmann) [inline bpf/lib/policy.h]: This looks not correct. You are loading icmp hdr from the packet and converting it with htons. If so it should say bpf_ntohs(), but why is it even needed for u8 load?

File: bpf/lib/policy.h [hunk 2]
```diff
@@ -87,13 +81,7 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 local_id,
...
 				return DROP_INVALID;
 
-			/* Convert from unsigned char to unsigned short
-			 * considering byte order(little-endian).
-			 * In the little-endian case, for example, 2byte data "AB"
-			 * convert to "BA".
-			 * Therefore, the "icmp_type" should be shifted not just casting.
-			 */
-			key.dport = (__u16)(icmp_type << 8);
+			key.dport = bpf_htons(icmp_type);
```
Comment (borkmann) [inline bpf/lib/policy.h]: Why is this needed?

File: bpf/lib/policy.h (hunk 1 above)
Comment (joestringer) [reply bpf/lib/policy.h]: I'm also a bit confused. I'm not sure if it's the comment that's incorrect or the code. The key to understanding this will be to confirm exactly where in the dport field the ICMP type is populated by userspace and ensure the datapath matches this.

I'm wondering if it's just the comment that's wrong, which is to say that the encoding of the ICMP type into the dport field has nothing to do with endianness, it's just that the "upper 8 bits" of the 16-bit word is where the ICMP type lives, so the code must do a bitshift.

Comment (joestringer) [reply bpf/lib/policy.h]: Looks like the ICMP type is parsed in and held in userspace in the host byte-order 16-bit dport field:

https://github.com/cilium/cilium/blob/687e4f8f403aa75f3e1e42005db51b29fdd9e5ef/pkg/policy/api/icmp.go#L208-L211
...
https://github.com/cilium/cilium/blob/687e4f8f403aa75f3e1e42005db51b29fdd9e5ef/pkg/policy/l4.go#L772

Which then gets converted to network byte-order (big-endian) when pushed into the datapath:

https://github.com/cilium/cilium/blob/687e4f8f403aa75f3e1e42005db51b29fdd9e5ef/pkg/maps/policymap/policymap.go#L279

If we consider the in-memory layout of the network byte-order/NBO/big-endian dport field for a policy applying to ICMP type `XX`, we'd expect it to be `00 XX`.

So what we need is that the u8 gets extended to u16 but with (a) big endian architecture: lower 8 bits set to the type value and (b) little endian architecture: upper 8 bits set to the type value.

I agree it's a bit misleading to do this byteswap with `bpf_htons()` given that the value does not start in host byte-order (how could it? It's only 8 bits), but likewise it's a bit misleading to byteswap with `bpf_ntohs()` because the target is not host byte-order. If we break this down, the u8 gets copied into a host byte order u16 variable, so doing `htons()` is more accurate?

Comment (chez-shanpu) [reply bpf/lib/policy.h]: Thank you for your review!

> so doing htons() is more accurate?

Or should we keep the code and fix the comments?
As you say, using `bpf_htons()` may confuse other developers.

Comment (joestringer) [reply bpf/lib/policy.h]: I think that keeping `<< 8` is incorrect on big-endian systems, because the value in userspace is stored in big-endian with the lowest byte populated, transferred to the datapath with a no-op byteswap because the system is big-endian, so you must perform a lookup with a big-endian u16 with the lower byte populated. If you read the 8-bit type field into a big-endian u16 then shift it up by 8 bits, you will end up performing a lookup with the _upper_ 8 bits populated. The lookup would then fail.

Comment (joestringer) [reply bpf/lib/policy.h]: Assuming that we confirm my reasoning above, an alternative could be to just define a new macro in `bpf/lib/endian.h` something like the below, and use that instead:

```
#define bpf_u8_to_be16(x) bpf_htons(x)
```

With this and a well-worded commit message, I think it should be fairly clear for future developers.

Comment (chez-shanpu) [reply bpf/lib/policy.h]: I see. It seems that we have to use something like `htons()` not shift.

> define a new macro in bpf/lib/endian.h something like the below, and use that instead

It looks nice to me!

Comment (kaworu) [reply bpf/lib/policy.h]: Throwing another suggestion:
```suggestion
			key.dport = bpf_htons((__u16)icmphdr.type);
```
Casting explicitly instead of relying on the implicit conversion convey the intent better IMHO.

Comment (joestringer) [reply bpf/lib/policy.h]: Yeah cast looks a bit more explicit :+1:

File: bpf/lib/policy.h [hunk 3]
```diff
@@ -69,13 +69,7 @@ __policy_can_access(const void *map, struct __ctx_buff *ctx, __u32 local_id,
...
 
 # if defined(ENABLE_ICMP_RULE)
-			/* Convert from unsigned char to unsigned short
-			 * considering byte order(little-endian).
-			 * In the little-endian case, for example, 2byte data "AB"
-			 * convert to "BA".
-			 * Therefore, the "icmp_type" should be shifted not just casting.
-			 */
-			key.dport = (__u16)(icmphdr.type << 8);
+			key.dport = bpf_u8_to_be16((__u16)icmphdr.type);
```
Comment (ldelossa) [inline bpf/lib/policy.h]: I don't mean to bike shed this too much, but now that you macro'd away the call to `bpf_htons` might it make sense to place the (__u16) cast **in** that macro? The macro is semantically saying "we'll do the cast for you" IMO.

**Review by ldelossa — APPROVED** [review]
This looks good to me now. Just a small nit I made.

File: bpf/lib/policy.h (hunk 3 above)
Comment (chez-shanpu) [reply bpf/lib/policy.h]: Thank you for your review!
I fixed it :)

**Review by joestringer — APPROVED** [review]

**Comment by joestringer** [comment]
Thanks for the fix @chez-shanpu !
